// mutt-config-core/index.js
/**
 * mutt-config-core
 * Pure configuration generator for Neomutt + Gmail + OAuth2.0
 *
 * No dependencies. No side effects. Just pure functions.
 */

/**
 * @typedef {Object} MuttConfigInput
 * @property {string} email - Gmail address
 * @property {string} realName - User's real name
 * @property {string} editor - Editor command (e.g., 'nvim', 'vim')
 * @property {'ja'|'en'} locale - Gmail language setting
 * @property {string} [gpgKeyId] - GPG key ID (defaults to email)
 */

/**
 * @typedef {Object} MuttConfigOutput
 * @property {string} accountMuttrc - Account-specific configuration
 * @property {string} mainMuttrc - Main muttrc file
 * @property {string} setupGuide - Next steps guide for the user
 */

/**
 * Gmail folder names by locale
 */
const GMAIL_FOLDERS = {
  ja: {
    drafts: "下書き",
    sent: "送信済みメール",
    trash: "ゴミ箱",
    allMail: "すべてのメール",
  },
  en: {
    drafts: "Drafts",
    sent: "Sent Mail",
    trash: "Trash",
    allMail: "All Mail",
  },
};

/**
 * Validate input configuration
 * @param {MuttConfigInput} input
 * @throws {Error} If validation fails
 */
function validateInput(input) {
  if (!input.email || !/^[^\s@]+@gmail\.com$/.test(input.email)) {
    throw new Error("Invalid Gmail address");
  }

  if (!input.realName || input.realName.trim().length === 0) {
    throw new Error("Real name is required");
  }

  if (!input.editor || input.editor.trim().length === 0) {
    throw new Error("Editor is required");
  }

  if (!["ja", "en"].includes(input.locale)) {
    throw new Error('Locale must be "ja" or "en"');
  }
}

/**
 * Generate account-specific muttrc configuration
 * @param {MuttConfigInput} input
 * @returns {string}
 */
function generateAccountMuttrc(input) {
  const folders = GMAIL_FOLDERS[input.locale];
  const gpgKeyId = input.gpgKeyId || input.email;

  return `# Gmail account settings for ${input.email}
# Generated by mutt-config-core

set realname = "${input.realName}"
set from = "${input.email}"
set smtp_url = "smtps://${input.email}@smtp.gmail.com:465/"
set imap_user = "${input.email}"
set folder = "imaps://imap.gmail.com:993"
set spoolfile = "+INBOX"
set postponed = "+[Gmail]/${folders.drafts}"
set record = "+[Gmail]/${folders.sent}"
set trash = "+[Gmail]/${folders.trash}"

# Auto-detect mailboxes
set imap_check_subscribed = yes

# Performance optimizations
set imap_passive = no
set imap_keepalive = 300
set imap_idle = yes
set mail_check = 60
set timeout = 10

# Cache settings
set header_cache = "~/.cache/mutt/headers"
set message_cachedir = "~/.cache/mutt/bodies"
set imap_headers = "X-PRIORITY X-SPAM-STATUS"

# OAuth 2.0 authentication (GPG encrypted)
set imap_authenticators = "oauthbearer:xoauth2"
set imap_oauth_refresh_command = "python3 ~/.config/mutt/mutt_oauth2.py --decryption-pipe 'gpg -d' ~/.local/etc/oauth-tokens/gmail.tokens"
set smtp_authenticators = "oauthbearer:xoauth2"
set smtp_oauth_refresh_command = "python3 ~/.config/mutt/mutt_oauth2.py --decryption-pipe 'gpg -d' ~/.local/etc/oauth-tokens/gmail.tokens"

# Editor
set editor = "${input.editor}"
`;
}

/**
 * Generate main muttrc configuration
 * @param {MuttConfigInput} input
 * @returns {string}
 */
function generateMainMuttrc(input) {
  const folders = GMAIL_FOLDERS[input.locale];

  return `# vim: filetype=neomuttrc
# Main Neomutt configuration
# Generated by mutt-config-core

source /usr/local/share/mutt-wizard/mutt-wizard.muttrc

# Editor setting
set editor = "${input.editor}"

# Default account (Gmail)
source ~/.config/mutt/accounts/${input.email}.muttrc

# Sidebar settings (disabled - use macros instead)
set sidebar_visible = no

# Folder switching macros
macro index gi "<change-folder>=INBOX<enter>" "Go to inbox"
macro index gt "<change-folder>=+[Gmail]/${folders.trash}<enter>" "Go to trash"
macro index gs "<change-folder>=+[Gmail]/${folders.sent}<enter>" "Go to sent"
macro index gd "<change-folder>=+[Gmail]/${folders.drafts}<enter>" "Go to drafts"
macro index ga "<change-folder>=+[Gmail]/${folders.allMail}<enter>" "Go to all mail"
`;
}

/**
 * Generate setup guide with next steps
 * @param {MuttConfigInput} input
 * @returns {string}
 */
function generateSetupGuide(input) {
  const gpgKeyId = input.gpgKeyId || input.email;

  return `
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  Neomutt + Gmail Setup - Next Steps
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

Configuration files have been generated!

Next steps:

1. Download mutt_oauth2.py
   cd ~/.config/mutt/
   curl -O https://raw.githubusercontent.com/neomutt/neomutt/main/contrib/oauth2/mutt_oauth2.py
   chmod +x mutt_oauth2.py

2. Setup Google Cloud Console
   - Visit https://console.cloud.google.com/
   - Create a new project
   - Enable Gmail API
   - Configure OAuth consent screen (User Type: External)
   - Add scope: https://mail.google.com/
   - Add test user: ${input.email}
   - Create OAuth client ID (Desktop application)
   - Note down Client ID and Client Secret

3. Obtain OAuth 2.0 token
   python3 ~/.config/mutt/mutt_oauth2.py \\
     --authorize \\
     --authflow localhostauthcode \\
     --encryption-pipe 'gpg --encrypt --recipient ${gpgKeyId}' \\
     --client-id YOUR_CLIENT_ID.apps.googleusercontent.com \\
     --client-secret YOUR_CLIENT_SECRET \\
     --provider google \\
     --email ${input.email} \\
     ~/.local/etc/oauth-tokens/gmail.tokens

   Replace YOUR_CLIENT_ID and YOUR_CLIENT_SECRET with actual values.
   You will be prompted for your GPG passphrase.

4. Launch Neomutt
   neomutt

For detailed instructions, visit:
https://github.com/a-lost-social-misfit/terminal-blog

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
`;
}

/**
 * Generate all Neomutt configuration files
 * @param {MuttConfigInput} input - Configuration input
 * @returns {MuttConfigOutput} Generated configuration content
 * @throws {Error} If input validation fails
 */
function generateMuttConfigs(input) {
  validateInput(input);

  return {
    accountMuttrc: generateAccountMuttrc(input),
    mainMuttrc: generateMainMuttrc(input),
    setupGuide: generateSetupGuide(input),
  };
}

/**
 * Get file paths for configuration files
 * @param {string} email - Gmail address
 * @returns {Object} File paths (relative to home directory)
 */
function getConfigPaths(email) {
  return {
    accountMuttrc: `.config/mutt/accounts/${email}.muttrc`,
    mainMuttrc: ".config/mutt/muttrc",
    oauthTokens: ".local/etc/oauth-tokens",
    cacheHeaders: ".cache/mutt/headers",
    cacheBodies: ".cache/mutt/bodies",
  };
}

// Export main API
module.exports = {
  generateMuttConfigs,
  getConfigPaths,
  GMAIL_FOLDERS,
};

// For testing purposes
module.exports._internal = {
  validateInput,
  generateAccountMuttrc,
  generateMainMuttrc,
  generateSetupGuide,
};
